### PR TITLE
feat: implement namespace alias

### PR DESCRIPTION
**Pull request by maan2003**
Fixes 

https://github.com/carbon-language/carbon-lang/blob/46503c0a9d20d8e38dde1ff4266f4e2b98699669/explorer/interpreter/resolve_names.cpp#L281

This was super fun to implement as first contribution. Learnt a lot about how the interpreter works while digging through the code and trying different ways to fix it.

### CODE REVIEW

File: explorer/testdata/namespace/namespace_alias.carbon
Comment (maan2003) [inline explorer/testdata/namespace/namespace_alias.carbon]: how to deal with test segregation? Should split this into multiple tests?

File: explorer/interpreter/resolve_names.cpp
Comment (jonmeow) [inline explorer/interpreter/resolve_names.cpp]: Thanks! Can you change this to use `alias.target()` directly here, removing `namespace_aliases_`? That way, we can using the existing tracking instead of adding overlapping storage for aliased namespaces.

**Review by jonmeow — DISMISSED** [review]

File: explorer/testdata/namespace/namespace_alias.carbon
Comment (jonmeow) [inline explorer/testdata/namespace/namespace_alias.carbon]: This should probably use a different name than `F` so that it's more obvious if there's a lookup issue. (versus `N.F`)

File: explorer/testdata/namespace/namespace_alias.carbon
Comment (jonmeow) [reply explorer/testdata/namespace/namespace_alias.carbon]: This seems fine to me, the tests are sufficiently related and really only checking lookup validity.

**Comment by jonmeow** [comment]
Please also note that the test seems to be failing (https://github.com/carbon-language/carbon-lang/actions/runs/4547451617/jobs/8017343855?pr=2720)

File: explorer/interpreter/resolve_names.cpp
Comment (maan2003) [reply explorer/interpreter/resolve_names.cpp]: Do you mean moving `ResolveQualifier()` here?

Can't use `alias.target()` directly because it is of  type `Expression` (not a super type of `NamespaceDeclaration`)?

**Comment by maan2003** [comment]
> Please also note that the test seems to be failing (https://github.com/carbon-language/carbon-lang/actions/runs/4547451617/jobs/8017343855?pr=2720)

should be fixed now. just forgot to remove `%{not}`

File: explorer/interpreter/resolve_names.cpp
Comment (jonmeow) [reply explorer/interpreter/resolve_names.cpp]: > Do you mean moving `ResolveQualifier()` here?

I wouldn't expect it to, since the alias should already be resolved before reaching this.

> Can't use `alias.target()` directly because it is of type `Expression` (not a super type of `NamespaceDeclaration`)?

You're right, this may involve changes to AliasDeclaration to store its target as an AstNode instead of an Expression. However, it offers greater consistency to store aliased namespaces in the same way as other aliases; aliases are intended to support a wide variety of targets.

Comment (maan2003) [reply explorer/interpreter/resolve_names.cpp]: changing `target` to `AstNode*` doesn't work because we only have `const AstNode*` (ValueNodeView). 

we can't change it to `const AstNode*` because we need `Expression*` from `AliasDeclaration` sometimes.

Added `AliasDeclaration::resolved_declaration`, and found similar pattern is also used in other places. (eg: `ImplDeclaration::match_first`)

File: explorer/interpreter/type_checker.cpp
Comment (jonmeow) [inline explorer/interpreter/type_checker.cpp:5946]: The special-casing in type_checker.cpp makes me wonder what would happen if you try using an aliased namespace in a place that's not expecting an aliased namespace. e.g., add a test for `namespace N; alias A = N; fn Main() -> i32 { return A; }` and see what happens.

File: explorer/interpreter/resolve_names.cpp
Comment (jonmeow) [inline explorer/interpreter/resolve_names.cpp]: Does the invalid case of this have tests? e.g.:
- `var x: i32; alias A = x; fn A.Foo() {}`
- `var x: i32; alias A = x; fn Main() -> i32 { return A.b; }`

File: explorer/interpreter/resolve_names.cpp
Comment (jonmeow) [reply explorer/interpreter/resolve_names.cpp]: Thanks, I think I didn't look closely enough. I'm still trying to understand the current code, though. It looks like `resolved_declaration` is going to be `alias.target().base()`. Why doesn't the code in SimpleMemberAccessExpression use `alias.target().base()` directly instead of adding the new `resolved_declaration`?

**Review by jonmeow — COMMENTED** [review]
I'm going to have to step back from this review for a week or two because I'm going to be busy. @zygoloid can you pick this up if there's progress while I'm out?

File: explorer/interpreter/resolve_names.cpp
Comment (maan2003) [reply explorer/interpreter/resolve_names.cpp]: > It looks like resolved_declaration is going to be alias.target().base().

No, alias->target() is just an expression, and resolve_decl is the ast node of declaration of the namespace.

running with 
```
llvm::outs() << "resolved_decl = " << *resolved_decl.value() << " target = " << alias.target() << "\n";
```
resolved_decl = namespace N; target = N

File: explorer/interpreter/type_checker.cpp
Comment (zygoloid) [inline explorer/interpreter/type_checker.cpp:5940]: Do we need this special case? It would seem simpler to unconditionally set the constant value here if it's not doing any harm, and I think there are still cases where we'll look at this and assert / crash if it's not set.

**Review by zygoloid — COMMENTED** [review]
I think the code change here is in good shape, but I would like to see the additional testing that @jonmeow requested, for invalid cases such as:

- `namespace N; alias A = N; fn Main() -> i32 { return A; }`
- `var x: i32; alias A = x; fn A.Foo() {}`
- `var x: i32; alias A = x; fn Main() -> i32 { return A.b; }`

**Comment by maan2003** [comment]
> `var x: i32; alias A = x; fn A.Foo() {}`

var aliases are not allowed. But I have added tests for interface alias.

**Comment by maan2003** [comment]
I am not sure why git is showing:
```
similarity index 58%
rename from explorer/testdata/namespace/fail_namespace_alias.carbon
rename to explorer/testdata/alias/fail_wrong_interface_alias_use.carbon
```

I have no such commit. I will try to fix this when I squash all the commits at the end. Please don't merge without this.

File: explorer/interpreter/type_checker.cpp
Comment (maan2003) [reply explorer/interpreter/type_checker.cpp:5940]: otherwise `InterpExt(alias->target)` fails with the error like indicated in the removed fail_namespace_alias.test.  
>  could not find `namespace N;

This is because namespace declaration does not have constant value, just a static_value.

https://github.com/carbon-language/carbon-lang/blob/0dc80bac01acc21e15e72cf4ac060efa8a61bb59/explorer/interpreter/type_checker.cpp#L6072-L6076

I think this is because all namespace related expressions are replaced during name resolution.

> I think there are still cases where we'll look at this and assert / crash if it's not set.

This should because similar to namspace declaration, which also doesn't have a constant value.

**Comment by jonmeow** [comment]
> > `var x: i32; alias A = x; fn A.Foo() {}`
> 
> var aliases are not allowed. But I have added tests for interface alias.

Fair enough. @zygoloid is it your expectation that aliases of variables shouldn't be supported? This doesn't match mine, and I worry that an approach excluding variables right now is going to lead to bugs.

Regardless, in my examples, it should be equivalent to replace the `var x` with some other aliasable thing, e.g.:

```
fn F() {}
alias A = F;
fn A.Foo() {}
```

**Comment by maan2003** [comment]
I have added more tests for this.

**Comment by zygoloid** [comment]
> I am not sure why git is showing:
> 
> ```
> similarity index 58%
> rename from explorer/testdata/namespace/fail_namespace_alias.carbon
> rename to explorer/testdata/alias/fail_wrong_interface_alias_use.carbon
> ```
> 
> I have no such commit.

Don't worry about this. `git` doesn't actually track file renames or copies, it just checks to see whether an added file looks a lot like a removed file, in which case it concludes that the file was renamed. Because most of a test file is the copyright block, the `RUN` lines, and the definition of `Main`, it often happens that `git` incorrectly concludes that a rename was performed when tests are added and removed in the same commit.

File: explorer/interpreter/resolve_names.cpp
Comment (jonmeow) [inline explorer/interpreter/resolve_names.cpp]:
```suggestion
      if (const auto* base_namespace =
              dyn_cast<NamespaceDeclaration>(&scope->base())) {
        namespace_decl = base_namespace;
      } else if (const auto* alias = dyn_cast<AliasDeclaration>(&scope->base());
                 alias && alias->resolved_node()) {
        if (const auto* alias_namespace =
                dyn_cast<NamespaceDeclaration>(*alias->resolved_node())) {
          namespace_decl = alias_namespace;
        }
      }
```

Let's make more use of dyn_cast here instead of isa+cast

File: explorer/interpreter/resolve_names.cpp
Comment (jonmeow) [inline explorer/interpreter/resolve_names.cpp]:
```suggestion
        auto ns_it = namespace_scopes_.find(*namespace_decl);
```

I believe the codebase prefers `*` over `.value()` on optionals.

File: explorer/interpreter/resolve_names.cpp
Comment (jonmeow) [reply explorer/interpreter/resolve_names.cpp]: Thanks for the clarification. I was out, but I've taken some more time to look at the implementation now and to understand a bit better what's going on.

**Review by jonmeow — COMMENTED** [review]
Sorry about the delay -- I was out for a week, and now zygoloid is out for a week.

Looking over the discussion on type_checker regarding the special-casing of namespaces, it may be reasonable to do something like that, but it made me realize indirect aliases aren't being addressed. Consider:

```
namespace N;

alias A = N;
alias B = A;
alias C = B;

fn N.F() {}

fn Main() -> i32 {
  C.F();
  return 0;
}
```

This example fails because the member access code currently doesn't look through indirect aliases.

Similarly, does the type checker code fail if it the alias points at an alias? Because of the above error, I'm not seeing it get that far, but it's probably something to look at and think about.

File: explorer/testdata/namespace/namespace_alias.carbon
Comment (maan2003) [inline explorer/testdata/namespace/namespace_alias.carbon:1]: should this be `alias/namespace_alias.carbon`?

File: explorer/interpreter/resolve_names.cpp
Comment (maan2003) [inline explorer/interpreter/resolve_names.cpp]: this feels nicer to me.

but an alternative:

```cpp
while (const auto* alias = dyn_cast<AliasDeclaration>(base)) {
  if (auto resolved = alias->resolved_declaration()) {
    base = *resolved;
  } else {
    break;
  }
}
```

Comment (jonmeow) [reply explorer/interpreter/resolve_names.cpp]: I'd prefer the alternative approach; the use of a lambda adds more flow control paths to understand, whereas the `while`/`break` approach seems simpler. But also, looking at the alternative I see a simplification that I've suggested (separately, so that I can suggest the full replacement).

Comment (jonmeow) [inline explorer/interpreter/resolve_names.cpp]:
```suggestion
      // Recursively resolve aliases.
      while (const auto* alias = dyn_cast<AliasDeclaration>(base); alias && alias->resolved_declaration()) {
        base = *alias->resolved_declaration();
      }
```

File: explorer/testdata/namespace/namespace_alias.carbon
Comment (jonmeow) [reply explorer/testdata/namespace/namespace_alias.carbon:1]: I don't have a strong opinion, but I think it should be next to namespace_alias_recursive, so probably `alias/` makes sense (or you could move the other namespace alias tests here).

**Review by jonmeow — APPROVED** [review]
Noting, minor change. Once the final edits are in I'll merge, I think it's straightforward.

File: explorer/interpreter/resolve_names.cpp
Comment (maan2003) [reply explorer/interpreter/resolve_names.cpp]: this was also my first approach. But looks like c++ doesn't support while with initializer. 
closest is for without update, but in that case control flow is incorrect because init is run only once not every iteration.

Comment (maan2003) [reply explorer/interpreter/resolve_names.cpp]: I have switched to the alternative.

**Review by jonmeow — APPROVED** [review]

File: explorer/interpreter/resolve_names.cpp
Comment (jonmeow) [reply explorer/interpreter/resolve_names.cpp]: Right, and I think I've even seen the proposal for `while`...

**Comment by jonmeow** [comment]
I think alias/fail_namespace_alias_member.carbon needs to be updated due to the move

File: explorer/interpreter/resolve_names.cpp
Comment (jonmeow) [reply explorer/interpreter/resolve_names.cpp]: What am I thinking:

for (const auto* alias = dyn_cast<AliasDeclaration>(base); alias && alias->resolved_declaration(); dyn_cast<AliasDeclaration>(base)) {

(although I suppose that's a little redundant -- don't do this)

**Pull request merged**
